pkcs11/src/new/mod.rs

```diff
@@ -35,20 +47,88 @@ impl Pkcs11 {
                 pkcs11_sys::Pkcs11::new(filename.as_ref()).map_err(Error::LibraryLoading)?;
             let mut list = mem::MaybeUninit::uninit();
 
-            if pkcs11_lib.can_call().C_GetFunctionList().is_err() {
-                return Err(Error::LibraryLoading(libloading::Error::DlOpenUnknown));
-            }
-
             Rv::from(pkcs11_lib.C_GetFunctionList(list.as_mut_ptr())).into_result()?;
 
             let list_ptr = *list.as_ptr();
 
             Ok(Pkcs11 {
                 _pkcs11_lib: pkcs11_lib,
                 function_list: *list_ptr,
+                logged_sessions: Mutex::new(HashMap::new()),
+                pins: RwLock::new(HashMap::new()),
             })
         }
     }
+
+    pub fn set_pin(&self, slot: Slot, pin: &str) -> Result<()> {
+        let _ = self
+            .pins
+            .write()
+            .expect("Pins lock poisoned")
+            .insert(slot, Secret::new(CString::new(pin)?.into_bytes()));
+        Ok(())
+    }
+
+    // Do not fail if the user is already logged in. It happens if another session on the same slot
+    // has already called the log in operation. Record the login call and only log out when there
+    // aren't anymore sessions requiring log in state.
+    fn login(&self, session: &Session, user_type: UserType) -> Result<()> {
+        let pins = self.pins.read().expect("Pins lock poisoned");
+        let pin = pins
+            .get(&session.slot())
+            .ok_or(Error::PinNotSet)?
+            .expose_secret();
+
+        let mut logged_sessions = self
+            .logged_sessions
+            .lock()
+            .expect("Logged sessions mutex poisoned!");
+
+        match unsafe {
+            Rv::from(get_pkcs11!(self, C_Login)(
```

Comment:
Btw, I've not looked at all the uses of this, but if `get_pkcs11` always takes `self` as the first argument, you could just use `self` in the macro without having to pass it (I think?)

Reply:
I tried in the beginning when I created this macro but it's impossible to easily do that 😢 See [this](https://stackoverflow.com/questions/44120455/how-to-call-methods-on-self-in-macros) for example

Reply:
Ok, makes sense!